### ISSUE TITLE
`hide` does not fully hide modules

### ISSUE THREAD

**Issue by amtoine** [nushell/nushell#8281]
### Describe the bug

i was trying to hide a module but couldn't get it to fully be removed.

**TL;DR**: it is removed but still accessible with `help modules | where name == <module>`

### functions
- we have no `foo` function in scope
```bash
>_ help foo
Error: nu::parser::not_found

  × Not found.
   ╭─[entry #7:1:1]
 1 │ help foo
   ·      ─┬─
   ·       ╰── did not find anything under this name
   ╰────
```
- let's define `foo`
```bash
>_ def foo [] { "this is foo" }
```
- and use it
```bash
>_ help foo
Usage:
  > foo

Flags:
  -h, --help - Display the help message for this command


>_ foo
this is foo
```
- now we hide it
```
>_ hide foo
>_ foo
Error: nu::shell::external_command

  × External command failed
   ╭─[entry #12:1:1]
 1 │ foo
   · ─┬─
   ·  ╰── did you mean 'for'?
   ╰────
  help: No such file or directory (os error 2)

>_ help foo
Error: nu::parser::not_found

  × Not found.
   ╭─[entry #13:1:1]
 1 │ help foo
   ·      ─┬─
   ·       ╰── did not find anything under this name
   ╰────
```
:heavy_check_mark: all is fine

### aliasea
- we have no `bar` alias in scope
>_ help bar
Error: nu::parser::not_found

  × Not found.
   ╭─[entry #14:1:1]
 1 │ help bar
   ·      ─┬─
   ·       ╰── did not find anything under this name
   ╰────
```
- let's define it
```bash
>_ alias bar = echo "this is bar"
```
- and use it a bit
```bash
>_ bar
this is bar
>_ help bar | lines | first  5
Returns its arguments, ignoring the piped-in value.

When given no arguments, it returns an empty string. When given one argument,
it returns it. Otherwise, it returns a list of the arguments. There is usually
little reason to use this over just writing the values as-is.
>_ bar "foo"
╭───┬─────────────╮
│ 0 │ this is bar │
│ 1 │ foo         │
╰───┴─────────────╯
```
- now hide it back
```bash
>_ hide bar
>_ bar
Error: nu::shell::external_command

  × External command failed
   ╭─[entry #20:1:1]
 1 │ bar
   · ─┬─
   ·  ╰── did you mean 'br'?
   ╰────
  help: No such file or directory (os error 2)

>_ help bar
Error: nu::parser::not_found

  × Not found.
   ╭─[entry #21:1:1]
 1 │ help bar
   ·      ─┬─
   ·       ╰── did not find anything under this name
   ╰────
```
:heavy_check_mark: all is fine

### modules
- we do not have a `baz` module right now
```bash
>_ help "baz fn"
Error: nu::parser::not_found

  × Not found.
   ╭─[entry #23:1:1]
 1 │ help "baz fn"
   ·      ────┬───
   ·          ╰── did not find anything under this name
   ╰────
```
- let's define it and use it
```bash
>_ module baz { export def fn [] { "this is baz::fn" } }
>_ use baz
>_ help "baz fn"
Usage:
  > baz fn

Flags:
  -h, --help - Display the help message for this command


>_ baz fn
this is baz::fn
```
- and hide it
```bash
>_ hide baz
>_ baz fn
Error: nu::shell::external_command

  × External command failed
   ╭─[entry #28:1:1]
 1 │ baz fn
   · ─┬─
   ·  ╰── executable was not found
   ╰────
  help: No such file or directory (os error 2)

>_ help baz fn
Error: nu::parser::not_found

  × Not found.
   ╭─[entry #30:1:1]
 1 │ help baz fn
   ·      ───┬──
   ·         ╰── did not find anything under this name
   ╰────

>_ help "baz fn"
Error: nu::parser::not_found

  × Not found.
   ╭─[entry #31:1:1]
 1 │ help "baz fn"
   ·      ────┬───
   ·          ╰── did not find anything under this name
   ╰────
```
:heavy_check_mark: all is fine

:x: but i see the following
```bash
>_ help baz
Module: baz

Exported commands:
  fn

This module does not export environment.
```

and here is what we get with plain `help`
```bash
>_ help commands | where name == foo
╭────────────╮
│ empty list │
╰────────────╯
>_ help aliases | where name == bar
╭────────────╮
│ empty list │
╰────────────╯
```
:x: but
```bash
>_ help modules | where name == baz
╭───┬──────┬────────────┬────────────────┬───────────┬───────╮
│ # │ name │  commands  │    aliases     │ env_block │ usage │
├───┼──────┼────────────┼────────────────┼───────────┼───────┤
│ 0 │ baz  │ ╭───┬────╮ │ [list 0 items] │           │       │
│   │      │ │ 0 │ fn │ │                │           │       │
│   │      │ ╰───┴────╯ │                │           │       │
╰───┴──────┴────────────┴────────────────┴───────────┴───────╯
```

### Expected behavior

i was expecting the following 
```bash
>_ help baz
Error: nu::parser::not_found

  × Not found.
   ╭─[entry #30:1:1]
 1 │ help baz
   ·      ─┬─
   ·       ╰── did not find anything under this name
   ╰────
```
and
```bash
>_ help modules | where name == baz
╭────────────╮
│ empty list │
╰────────────╯
```

### Configuration

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.76.1                                   |
| branch             | main                                     |
| commit_hash        | 0a1af852009d2f87dcee336ca61d5f41f4648cb0 |
| build_os           | linux-x86_64                             |
| build_target       | x86_64-unknown-linux-gnu                 |
| rust_version       | rustc 1.66.1 (90743e729 2023-01-10)      |
| rust_channel       | 1.66.1-x86_64-unknown-linux-gnu          |
| cargo_version      | cargo 1.66.1 (ad779e08b 2023-01-10)      |
| pkg_version        | 0.76.1                                   |
| build_time         | 2023-03-01 19:26:49 +01:00               |
| build_rust_channel | release                                  |
| features           | default, zip                             |
| installed_plugins  | gstat                                    |

**Comment by amtoine** [comment]
and also, `hide` does not appear to hide anything when used in `the `env_change.PWD` hook :thinking:

**Comment by kubouch** [comment]
`hide` hides only definitions brought in from the module, it does not attempt to hide the module itself. For environment variables, use `hide-env`. Hiding is also scoped, so whatever you hide in `env_change.PWD` hook will stay hidden only in the hook. To preserve the changes, you'd need to define the hook as a string.

**Comment by amtoine** [comment]
> `hide` hides only definitions brought in from the module, it does not attempt to hide the module itself.
>
mm, ok :thinking: 
so `baz` will never disappear?
is this what people are likely to expect? :confused: 

> For environment variables, use `hide-env`.
>
that makes sense :ok_hand: 

> Hiding is also scoped, so whatever you hide in `env_change.PWD` hook will stay hidden only in the hook. To preserve the changes, you'd need to define the hook as a string.
>
let me try this :yum:

**Comment by amtoine** [comment]
nice, that works :ok_hand:

**Comment by amtoine** [comment]
@kubouch 
so there is only this question about the module not being removed from `$nu.scope.modules` :+1:

**Comment by amtoine** [comment]
it appears the `baz` module goes into the `$nu.scope` as soon as it is defined
https://asciinema.org/a/564219
> **Note**
> the same exact behaviour by replacing all instances of `$nu.scope.modules` by `help modules`
